sonar-db/src/main/java/org/sonar/db/version/v564/CleanUsurperRootComponents.java

```diff
@@ -0,0 +1,156 @@
+    super(db);
+  }
+
+  @Override
+  public void execute(Context context) throws SQLException {
+    // fix snapshots which don't have the scope and/or qualifier of their associated component
+    fixSnapshotScopeAndQualifier(context);
+    // delete components declaring themselves as root in table PROJECTS but which don't have a root scope and/or qualifier
+    cleanUsurperRootComponents(context);
+    // components which has snapshots reference a component as root which is not a root
+    cleanSnapshotWithIncorrectRoot(context);
+  }
+
+  private static void fixSnapshotScopeAndQualifier(Context context) throws SQLException {
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select("select" +
+      " sn.id,p.scope,p.qualifier" +
+      " from" +
+      " snapshots sn, projects p" +
+      " where" +
+      " p.uuid = sn.component_uuid" +
+      " and (p.qualifier<>sn.qualifier or p.scope<>sn.scope)");
+    massUpdate.update("update snapshots set scope=?,qualifier=? where id=?");
+    massUpdate.rowPluralName("snapshots with inconsistent scope or qualifier");
+    massUpdate.execute((row, update) -> {
+      long snapshotId = row.getLong(1);
+      String scope = row.getString(2);
+      String qualifier = row.getString(3);
+
+      update.setString(1, scope);
+      update.setString(2, qualifier);
+      update.setLong(3, snapshotId);
+
+      return true;
+    });
+  }
+
+  private static void cleanUsurperRootComponents(Context context) throws SQLException {
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select("select p.id,p.uuid from projects p " +
+      " where" +
+      " p.project_uuid = p.uuid" +
+      " and not (" +
+      " p.scope = 'PRJ'" +
+      " and p.qualifier in ('TRK', 'VW', 'DEV')" +
+      " )");
+    massUpdate.update("delete from duplications_index where snapshot_id in (select id from snapshots where component_uuid=?)");
+    massUpdate.update("delete from project_measures where component_uuid=?");
+    massUpdate.update("delete from ce_activity where component_uuid=?");
+    massUpdate.update("delete from events where component_uuid=?");
+    massUpdate.update("delete from project_links where component_uuid=?");
+    massUpdate.update("delete from snapshots where component_uuid=? or root_component_uuid=?");
+    massUpdate.update("delete from issues where component_uuid=? or project_uuid=?");
+    massUpdate.update("delete from file_sources where file_uuid=? or project_uuid=?");
+    massUpdate.update("delete from group_roles where resource_id=?");
+    massUpdate.update("delete from user_roles where resource_id=?");
+    massUpdate.update("delete from properties where resource_id=?");
+    massUpdate.update("delete from widgets where resource_id=?");
+    massUpdate.update("delete from projects where uuid=?");
+    massUpdate.rowPluralName("usurper root components");
+    massUpdate.execute((row, update, updateIndex) -> {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this lambda expression number of lines from 30 to at most 20. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1188)